src/token.rs

```diff
@@ -662,80 +897,181 @@ impl Token {
-                match self.store_pin_object(
-                    USER_PIN_UID.to_string(),
-                    "User PIN".to_string(),
-                    pin.clone(),
-                ) {
-                    Ok(()) => (),
-                    Err(_) => return CKR_GENERAL_ERROR,
-                }
+    fn clear_private_session_objects(&mut self) {
+        let mut priv_handles = Vec::<CK_OBJECT_HANDLE>::new();
+        for (handle, obj) in self.session_objects.iter() {
+            if obj.is_private() {
+                priv_handles.push(*handle);
             }
-            CKU_SO => {
-                if old.is_none() {
-                    return CKR_PIN_INCORRECT;
-                }
-                let ret =
-                    self.so_login.change_pin(&self.info, pin, old.unwrap());
-                if ret != CKR_OK {
-                    return ret;
-                }
-                /* update pin in storage */
-                match self.store_pin_object(
-                    SO_PIN_UID.to_string(),
-                    "SO PIN".to_string(),
-                    pin.clone(),
-                ) {
-                    Ok(()) => (),
-                    Err(_) => return CKR_GENERAL_ERROR,
-                }
+        }
+
+        /* remove all private session objects */
+        for handle in priv_handles {
+            let _ = self.session_objects.remove(&handle);
+            self.handles.remove(handle);
+        }
+    }
+
+    fn clear_session_objects(&mut self, session: CK_SESSION_HANDLE) {
+        let mut handles: Vec<CK_OBJECT_HANDLE> = Vec::new();
+        for (_, obj) in self.session_objects.iter() {
+            if obj.get_session() == session {
+                handles.push(obj.get_handle());
             }
-            _ => return CKR_GENERAL_ERROR,
         }
 
-        /* If we set a PIN it means we switched to require Logins */
-        self.info.flags |= CKF_LOGIN_REQUIRED;
+        for h in handles {
+            let _ = self.session_objects.remove(&h);
+            self.handles.remove(h);
+        }
+    }
+
+    /* We should probably have lifetimes to ensure iv and aad are around for
+     * the lifetime of the returned structure, but this will require substantial
+     * reworking of the bindings, so for now we just get this comment.
+     * ENSURE the arguments stay in scope until CK_GCM_PARAMS is needed
+     * */
+    fn encryption_params(&self, iv: &[u8], aad: &[u8]) -> CK_GCM_PARAMS {
+        CK_GCM_PARAMS {
+            pIv: iv.as_ptr() as *mut CK_BYTE,
+            ulIvLen: iv.len() as CK_ULONG,
+            ulIvBits: (iv.len() * 8) as CK_ULONG,
+            pAAD: aad.as_ptr() as *mut CK_BYTE,
+            ulAADLen: aad.len() as CK_ULONG,
+            ulTagBits: 64 as CK_ULONG,
+        }
+    }
 
-        self.update_token_pin_flags();
+    fn encrypt_value(&self, uid: &String, val: &Vec<u8>) -> KResult<Vec<u8>> {
+        if let Some(ref kek) = self.kek {
+            let mut iv = [0u8; DEFAULT_IV_SIZE];
```

Comment:
I would use just `IV_SIZE` or `CRYPT_IV_SIZE` as I do not expect this will be somehow configurable in the long term, but its just a nit ...